man/mdoc.1
.\" 
.\" mdoc manual page.
.\" (C) 2008 Jonathan Pryor
.\" Author:
.\"   Jonathan Pryor (jpryor@novell.com)
.\"
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.TH "mdoc" 1
.SH NAME
mdoc \- Mono documentation management tool
.SH SYNOPSIS
.B mdoc
.I command
[\fIoptions\fR] [\fIargs\fR]
.SH OVERVIEW
\fImdoc\fR is an assembly-based documentation management system.
.PP
\fImdoc\fR permits creating and updating documentation \fIstubs\fR based on
the contents of an assembly.  It does not rely on documentation found within 
the source code.
.PP
The advantages are:
.TP
.B *
.B Code readability.
Good documentation is frequently (a) verbose, and (b)
filled with examples.  (For comparison, compare Microsoft .NET Framework
documentation, which is often a page or more of docs for each member, to
JavaDoc documentation, which can often be a sentence for each member.)
.Sp
Inserting good documentation into the source code can frequently bloat the
source file, as the documentation can be longer than the actual method that is
being documented.
.TP
.B *
.B Localization.
In-source documentation formats (such as \fIcsc /doc\fR)
have no support for multiple human languages.  If you need to support more
than one human language for documentation purposes, \fImdoc\fR
is useful as it permits each language's output to reside in its own directory, 
and \fImdoc\fR can add types/members for each separate documentation directory.
.TP
.B *
.B Administration.
It's not unusual to have separate documentation and development teams.  It's
also possible that the documentation team will have minimal experience with
the programming language being used.  In such circumstances, inline
documentation is not desirable as the documentation team could inadvertantly
insert an error into the source code while updating the documentation.
Alternatively, you may not want the documentation team to have access to the
source code for security reasons.  \fImdoc\fR allows the documentation to be 
kept \fIcompletely\fR separate and distinct from the source code used to 
create the assembly.
.PP
Documentation can be generated using the \fBmdoc update\fR command:
.nf

    mdoc update -o docs/en ProjectName.dll

.fi
Once the documentation stubs have been generated (and hopefully later filled 
in with actual documentation), there are three ways to view the documentation:
.TP
.B *
To generate a simple directory of HTML pages (one HTML file per type), use
\fBmdoc export-html\fR:
.nf

    mdoc export-html -o /srv/www/htdocs/ProjectName docs/en

.fi
.TP
.B *
To use an ASP.NET webapp to display the sources, see:
\fIhttp://anonsvn.mono-project.com/source/trunk/monodoc/engine/web/\fR.
.Sp
From a \fImonodoc\fR source checkout, you can do this:
.nf

    cd engine
    make web

.fi
This will use \fBxsp\fR(1) to serve the ASP.NET webapp; 
Visit \fIhttp://localhost:8080/\fR to view the documentation.
.TP
.B *
To use the \fBmonodoc\fR(1) documentation browser, you must first
\fIassemble\fR the documentation:
.nf

    mdoc assemble -o ProjectName docs/en

.fi
The above command creates the files \fIProjectName.tree\fR and
\fIProjectName.zip\fR.  An additional \fIProjectName.sources\fR file 
must be provided which describes where in the help system the documentation 
should be hooked up; it is a very simple XML file, like this:
.nf

    <?xml version="1.0"?>
    <monodoc>
      <source provider="ecma" basefile="ProjectName"
        path="various" />
    </monodoc>

.fi
The above configuration file describes that the documentation is in
ECMA format, that the base file name is \fIProjectName\fR and that it 
should be hooked up in the \fI"various"\fR part of the documentation tree.
If you want to look at the various nodes defined in the
documentation, you can look at the \fImonodoc.xml\fR file which is typically
installed in \fI/usr/lib/monodoc/monodoc.xml\fR.
.Sp
Once you have all of the required files (.zip, .tree and .sources) you can
install them into the system with the following command:
.nf

    cp ProjectName.tree ProjectName.zip ProjectName.source \\
      `pkg-config monodoc --variable sourcesdir`

.fi
The above will copy the files into the directory that Monodoc has
registered; you might need root permissions to do this.  The actual
directory is returned by the \fIpkg-config\fR invocation.
.SH MDOC COMMANDS
.PP
\fBmdoc assemble\fR
.RS 4
Compiles documentation for use within the \fBmonodoc\fR(1) browser.
.PP
See the \fBmdoc-assemble\fR(1) man page for details.
.RE
.PP
\fBmdoc export-html\fR
.RS 4
Exports documentation into a directory structure of HTML files.
.PP
See the \fBmdoc-export-html\fR(1) man page for details.
.RE
.PP
\fBmdoc export-msxdoc\fR
.RS 4
Exports documentation into the single-file 
\fIMicrosoft XML Documentation\fR format.
.PP
See the \fBmdoc-export-msxdoc\fR(1) man page for details.
.RE
.PP
\fBmdoc help\fR
.RS 4
View internal help for a given command.

.nf
    mdoc help assemble
.fi

is equivalent to:

.nf
    mdoc assemble --help
.fi

Multiple sub-commands may be listed at once:

.nf
    mdoc help assemble export-html update validate
.fi
.RE
.PP
\fBmdoc update\fR
.RS 4
Updates documentation, adding and removing members based upon a reference
assembly.
.PP
See the \fBmdoc-update\fR(1) man page for details.
.RE
.PP
\fBmdoc validate\fR
.RS 4
Validates the documentation against the Mono documentation schema.
.PP
See the \fBmdoc-validate\fR(1) man page for details.
.RE
.SH SEE ALSO
mdoc(5),
mdoc-assemble(1), 
mdoc-export-html(1), 
mdoc-update(1),
mdoc-validate(1)
.SH MAILING LISTS
.TP
Visit http://lists.ximian.com/mailman/listinfo/mono-docs-list for details.
.SH WEB SITE
Visit http://www.mono-project.com/docs/tools+libraries/tools/mdoc/ for details


NEWS
GIANT BEES DESCEND UPON BOSTON


mcs/tools/mdoc/Mono.Documentation/mdoc.cs
// Part of the mdoc(7) suite of tools.
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using Mono.Options;

namespace Mono.Documentation {

	class MDoc {

		private static bool debug;

		private static void Main (string[] args)
		{
			MDoc d = new MDoc ();
			try {
				d.Run (args);
			}
			catch (Exception e) {
				if (debug) {
					Console.Error.WriteLine ("mdoc: {0}", e.ToString ());
				}
				else {
					Console.Error.WriteLine ("mdoc: {0}", e.Message);

				}
				Console.Error.WriteLine ("See `mdoc help' for more information.");
				Environment.ExitCode = 1;
			}
		}

		int verbosity = 2;

		internal Dictionary<string, MDocCommand> subcommands;

		private void Run (string[] args)
		{
			subcommands = new Dictionary<string, MDocCommand> () {
				{ "assemble",         new MDocAssembler () },
				{ "dump-tree",        new MDocTreeDumper () },
				{ "export-html",      new MDocToHtmlConverter () },
				{ "export-html-webdoc",   new MDocExportWebdocHtml () },
				{ "export-msxdoc",    new MDocToMSXDocConverter () },
				{ "help",             new MDocHelpCommand (this) },
				{ "update",           new MDocUpdater () },
				{ "update-ecma-xml",  new MDocUpdateEcmaXml () },
				{ "validate",         new MDocValidator () },
				{ "index",            new MDocIndex () },
				{ "preserve",         new MDocPreserve () }
			};

			bool showVersion = false;
			bool showHelp    = false;
			var p = new OptionSet () {
				{ "version",  v => showVersion = v != null },
				{ "v:",       (int? v) => verbosity = v.HasValue ? v.Value : verbosity+1 },
				{ "debug",    v => debug = v != null },
				{ "h|?|help", v => showHelp = v != null },
				new ResponseFileSource (),
			};

			var extra = p.Parse (args);

			if (showVersion) {
				Console.WriteLine ("mdoc {0}", Consts.MonoVersion);
				return;
			}
			if (extra.Count == 0) {
				Console.WriteLine ("Use `mdoc help' for usage.");
				return;
			}
			if (showHelp) {
				extra.Add ("--help");
			}
			switch (extra [0]) {
				case "x-msitomsx":
					new MsidocToMsxdocConverter ().Run (extra);
					break;
				default: 
					GetCommand (extra [0]).Run (extra);
					break;
			}
		}

		internal MDocCommand GetCommand (string command)
		{
			MDocCommand h;
			if (!subcommands.TryGetValue (command, out h)) {
				Error ("Unknown command: {0}.", command);
			}
			h.TraceLevel  = (TraceLevel) verbosity;
			h.DebugOutput = debug;
			return h;
		}

		private static void Error (string format, params object[] args)
		{
			throw new Exception (string.Format (format, args));
		}
	}

	public abstract class MDocCommand {

		public TraceLevel TraceLevel { get; set; }
		public bool DebugOutput { get; set; }

		public abstract void Run (IEnumerable<string> args);

		protected List<string> Parse (OptionSet p, IEnumerable<string> args, 
				string command, string prototype, string description)
		{
			bool showHelp = false;
			p.Add ("h|?|help", 
					"Show this message and exit.", 
					v => showHelp = v != null );

			List<string> extra = null;
			if (args != null) {
				extra = p.Parse (args.Skip (1));
			}
			if (args == null || showHelp) {
				Console.WriteLine ("usage: mdoc {0} {1}", 
						args == null ? command : args.First(), prototype);
				Console.WriteLine ();
				Console.WriteLine (description);
				Console.WriteLine ();
				Console.WriteLine ("Available Options:");
				p.WriteOptionDescriptions (Console.Out);
				return null;
			}
			return extra;
		}

		public void Error (string format, params object[] args)
		{
			throw new Exception (string.Format (format, args));
		}

		public void Message (TraceLevel level, string format, params object[] args)
		{
			if ((int) level > (int) TraceLevel)
				return;
			if (level == TraceLevel.Error)
				Console.Error.WriteLine (format, args);
			else
				Console.WriteLine (format, args);
		}
	}

	class MDocHelpCommand : MDocCommand {

		MDoc instance;

		public MDocHelpCommand (MDoc instance)
		{
			this.instance = instance;
		}

		public override void Run (IEnumerable<string> args)
		{
			if (args != null && args.Count() > 1) {
				foreach (var arg in args.Skip (1)) {
					instance.GetCommand (arg).Run (new string[]{arg, "--help"});
				}
				return;
			}
			Message (TraceLevel.Warning, 
				"usage: mdoc COMMAND [OPTIONS]\n" +
				"Use `mdoc help COMMAND' for help on a specific command.\n" +
				"\n" + 
				"Available commands:\n\n   " +
				string.Join ("\n   ", instance.subcommands.Keys.OrderBy (v => v).ToArray()) +
				"\n\n" + 
				"mdoc is a tool for documentation management.\n" +
				"For additional information, see http://www.mono-project.com/"
			);
		}
	}
}

